### PR TITLE
Add String#matches_full, #match_full

### PR DESCRIPTION
**Pull request by straight-shoota**
Adds method to `String` to matche a regular expression over the entire string.

See #9309 for a related PR (which is probably not necessary after this one) and the discussion that brought up the idea for this.

### CODE REVIEW

File: spec/std/string_spec.cr [hunk 1]
```diff
@@ -2067,6 +2067,22 @@ describe "String" do
...
 
+  it "#matches_full?" do
+    "foo".matches_full?(/foo/).should be_true
+    "fooo".matches_full?(/foo/).should be_false
+    "ofoo".matches_full?(/foo/).should be_false
+    "pattern".matches_full?(/(\A)?pattern(\z)?/).should be_true
+    "_pattern_".matches_full?(/(\A)?pattern(\z)?/).should be_false
+  end
+
+  it "#match_full?" do
```
Comment (jhass) [inline spec/std/string_spec.cr:2078]:
```suggestion
  it "#match_full" do
```

File: src/string.cr [hunk 2]
```diff
@@ -4166,6 +4166,34 @@ class String
     regex.matches? self, pos
   end
 
+  # Returns `true` if the regular expression *regex* matches this string entirely.
```
Comment (jhass) [inline src/string.cr:4169]: Doesn't seem to return `true` ;)

**Comment by oprypin** [comment]
This implementation is important because it makes it easy to do the right thing.
That the implementation is simple is not relevant. In fact, maybe we'll find some vulnerability requiring to complicate it and then be able to fix it for all users going forward.
Some **wrong** things that this **already** prevents very well are:
1. Preventing `/\A#{expr}\z/`
    where `expr` can be `"foo|bar"`,
    making this `\Afoo` `|` `bar\z`
2. Preventing `/\A(#{expr})\z/`
    where `expr` can be `")|hax|("`,
    making this `\A()` `|` `hax` `|` `()\z`

**Comment by jhass** [comment]
Ruby handles this like this:
```
[32] pry(main)> /\A#{/a|b/}\z/
=> /\A(?-mix:a|b)\z/
```

Can we fix regex interpolation instead and cover a vastly broader amount of these kinds of missteps?

**Comment by jhass** [comment]
Actually  we do this properly already! https://carc.in/#/r/95qb

If you got a user supplied regex as a string you just have to pass it through  `Regex.new`.

File: src/string.cr [hunk 3]
```diff
@@ -4166,6 +4166,35 @@ class String
...
+  # It also updates `$~` with the result.
+  #
+  # ```
+  # "foo".match_full(/foo/)   # => Regex::MatchData("foo")
+  # $~                        # => Regex::MatchData("foo")
+  # "fooo".match_full(/foo/)  # => nil
+  # $~                        # raises Exception
+  # ```
+  def match_full(regex : Regex) : Regex::MatchData?
+    match = /(?:#{regex})\z/.match_at_byte_index(self, 0, Regex::Options::ANCHORED)
```
Comment (oprypin) [inline src/string.cr:4180]: Very interesting. So, according to https://github.com/crystal-lang/crystal/pull/9312#issuecomment-635180570, the brackets can even be dropped.
```suggestion
    match = /#{regex}\z/.match_at_byte_index(self, 0, Regex::Options::ANCHORED)
```

**Comment by oprypin** [comment]
@jhass Alright. But while that's true, the important part for making the wrong thing difficult is that the API accepts only `Regex`.

That `/\A#{/foo|bar/}\z/` works well doesn't mean that `/\A#{"foo|bar"}\z/` isn't still a problem.

Because the new API is the easy thing to do and it doesn't allow `String`, that alone can prevent vulnerabilities.

**Comment by jhass** [comment]
Well, I mean actually passing any user supplied string to a regex engine unchecked already is a vulnerability,  very easy to trigger a DoS like that. I'd like to make this harder, not easier :/

I think this API fixing one instance of the problem doesn't invalidate my argument. We cannot possibly provide a method for any single instance of this problem, so it's better to educate users on the general solution to it. And take the opportunity to emphasize that passing a user controlled string to the regex engine is rather dangerous.

Actually I wouldn't be too opposed to forbid interpolating strings into regex literals, you could always still build up the entire regex as a string instead if you really need  that, but it should tip most people towards `/#{Regex.new(...)}/`.

Summary is that I can't get rid of the feeling that this is fighting symptoms and not the problem.

**Comment by straight-shoota** [comment]
I actually never had in mind to promote this as a security feature. It's intended to be a simple helper for a relatively common functionality.
The security aspect is negligible because it doesn't add anything new.

The purpose is to provide a clean API for expressing the intent (making sure a regular expression matches against the entire string) instead of having to dip into details of implementation. 
If at some point we switch to a different regex engine that supports a full matching feature explicitly, these methods could be changed to use the appropriate library calls for that directly.

File: src/string.cr (hunk 3 above)
Comment (straight-shoota) [reply src/string.cr:4180]: Yeah, I am aware that they're not necessary. I guess they ended up here as result of a copy and paste.

**Comment by straight-shoota** [comment]
@jhass Removing String interpolation in Regex literals could be a useful security feature. But it's a different story.

**Comment by jhass** [comment]
Well, I approved this because I don't think it's a too terrible mistake. I just cannot really see the usecases personally.

**Comment by straight-shoota** [comment]
Oh, there are quite a few usecases for matching against a full string. Having a dedicated method for that feels better than having to dissect the meaning of the regular expression in order to understand it's supposed to match fully.

I personally tend to avoid regular expressions if possible, but I still have some direct for use cases for this:

https://github.com/shardbox/catalog/blob/0428d63ee84da594a61d07469358834b15a6a22e/src/awesome_list.cr#L76

https://github.com/shardbox/shardbox-core/blob/650dea420bfa7b4e8dfe88d7df2c0371609776b3/src/util/software_version.cr#L49

This is where the idea for #9309 was born:

https://github.com/straight-shoota/sanitize/blob/ea4f0256390a522438e22cf8842c19f3f365860c/src/policy/html_sanitizer.cr#L324-L326

**Comment by jhass** [comment]
I meant usecases where I don't see  `\A`/`\z` fit, which I find quite easy to understand and reason about.

**Comment by straight-shoota** [comment]
There's a related feature request for Ruby: [Introduce %R percent literal for anchored regular expression patterns
](https://bugs.ruby-lang.org/issues/18757). Might be worth considering for Crystal as well. But let's see what comes out of this for Ruby.

**Comment by straight-shoota** [comment]
I believe the same behaviour as proposed here can be achieved by passing the options `ANCHORED` and `ENDANCHORED` which requires the pattern to be anchored at the begin and end of the string, thus covering it fully. 
This would require no change to the API nor to regex patterns.
`ENDANCHORED` is only available in PCRE2.

**Pull request closed without merging**